Associated constants equality produces errors with constants

The following [code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fe20e502e384e7d4b1e92f2aba05187b) produces errors. These errors are both nonsensical and unexpected. This happens with `feature(associated_const_equality)` enabled or disabled, and both on nightly and stable.
```rust
#![feature(associated_const_equality)]

pub enum Mode {
    Cool,
}

pub trait Parse {
    const MODE: Mode;
}

pub trait CoolStuff: Parse<MODE = Mode::Cool> {}
```
Errors:
```
error[E0573]: expected type, found variant `Mode::Cool`
  --> src/lib.rs:11:35
   |
11 | pub trait CoolStuff: Parse<MODE = Mode::Cool> {}
   |                                   ^^^^^^^^^^
   |                                   |
   |                                   not a type
   |                                   help: try using the variant's enum: `crate::Mode`

error: expected associated constant bound, found type
  --> src/lib.rs:11:28
   |
11 | pub trait CoolStuff: Parse<MODE = Mode::Cool> {}
   |                            ^^^^^^^^^^^^^^^^^
   |
note: associated constant defined here
  --> src/lib.rs:8:5
   |
8  |     const MODE: Mode;
   |     ^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0573`.
```

These errors appear to happen whenever a constant value is used instead of a literal.

Another [example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ee9bd0786cd4bf5e1110cbfa6c587ee7) with constant variable instead of an enum variant.
```rust
#![feature(associated_const_equality)]

const MODE_COOL: bool = true;

pub trait Parse {
    const MODE: bool;
}

pub trait CoolStuff: Parse<MODE = MODE_COOL> {}
```

I am not going to post the error message again, as they are basically the same.

The current workaround that I have found is to wrap the constant value in `{}`, with will cause the code to compile on nightly with `feature(associated_const_equality)` enabled, or produce the `associated const equality is incomplete` error message otherwise.

### Meta
`rustc --version --verbose`:
```
rustc 1.68.0-nightly (0468a00ae 2022-12-17)
binary: rustc
commit-hash: 0468a00ae3fd6ef1a6a0f9eaf637d7aa9e604acc
commit-date: 2022-12-17
host: x86_64-pc-windows-msvc
release: 1.68.0-nightly
LLVM version: 15.0.6
```

--- a/compiler/rustc_hir_analysis/src/astconv/mod.rs
+++ b/compiler/rustc_hir_analysis/src/astconv/mod.rs
@@ -1199,17 +1199,26 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                     (_, _) => {
                         let got = if let Some(_) = term.ty() { "type" } else { "constant" };
                         let expected = def_kind.descr(assoc_item_def_id);
-                        let reported = tcx
-                            .sess
-                            .struct_span_err(
+                        let mut err = tcx.sess.struct_span_err(
+                            binding.span,
+                            &format!("expected {expected} bound, found {got}"),
+                        );
+                        err.span_note(
+                            tcx.def_span(assoc_item_def_id),
+                            &format!("{expected} defined here"),
+                        );
+
+                        if let hir::def::DefKind::AssocConst = def_kind
+                          && let Some(t) = term.ty() && (t.is_enum() || t.references_error())
+                          && tcx.features().associated_const_equality {
+                            err.span_suggestion(
                                 binding.span,
-                                &format!("expected {expected} bound, found {got}"),
-                            )
-                            .span_note(
-                                tcx.def_span(assoc_item_def_id),
-                                &format!("{expected} defined here"),
-                            )
-                            .emit();
+                                "if equating a const, try wrapping with braces",
+                                format!("{} = {{ const }}", binding.item_name),
+                                Applicability::HasPlaceholders,
+                            );
+                        }
+                        let reported = err.emit();
                         term = match def_kind {
                             hir::def::DefKind::AssocTy => {
                                 tcx.ty_error_with_guaranteed(reported).into()

--- a/src/test/ui/const-generics/assoc_const_eq_diagnostic.rs
+++ b/src/test/ui/const-generics/assoc_const_eq_diagnostic.rs
@@ -1,0 +1,18 @@
+#![feature(associated_const_equality)]
+
+pub enum Mode {
+    Cool,
+}
+
+pub trait Parse {
+    const MODE: Mode;
+}
+
+pub trait CoolStuff: Parse<MODE = Mode::Cool> {}
+//~^ ERROR expected associated constant bound
+//~| ERROR expected type
+
+fn no_help() -> Mode::Cool {}
+//~^ ERROR expected type, found variant
+
+fn main() {}

--- a/src/test/ui/const-generics/assoc_const_eq_diagnostic.stderr
+++ b/src/test/ui/const-generics/assoc_const_eq_diagnostic.stderr
@@ -1,0 +1,33 @@
+error[E0573]: expected type, found variant `Mode::Cool`
+  --> $DIR/assoc_const_eq_diagnostic.rs:11:35
+   |
+LL | pub trait CoolStuff: Parse<MODE = Mode::Cool> {}
+   |                                   ^^^^^^^^^^
+   |                                   |
+   |                                   not a type
+   |                                   help: try using the variant's enum: `Mode`
+
+error[E0573]: expected type, found variant `Mode::Cool`
+  --> $DIR/assoc_const_eq_diagnostic.rs:15:17
+   |
+LL | fn no_help() -> Mode::Cool {}
+   |                 ^^^^^^^^^^
+   |                 |
+   |                 not a type
+   |                 help: try using the variant's enum: `Mode`
+
+error: expected associated constant bound, found type
+  --> $DIR/assoc_const_eq_diagnostic.rs:11:28
+   |
+LL | pub trait CoolStuff: Parse<MODE = Mode::Cool> {}
+   |                            ^^^^^^^^^^^^^^^^^ help: if equating a const, try wrapping with braces: `MODE = { const }`
+   |
+note: associated constant defined here
+  --> $DIR/assoc_const_eq_diagnostic.rs:8:5
+   |
+LL |     const MODE: Mode;
+   |     ^^^^^^^^^^^^^^^^
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0573`.

